### PR TITLE
[feat] Add AMPnet implementation (clean version)

### PR DESCRIPTION
**Pull request by sidgoyal78**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
This PR adds the implementation of AMPnet (https://arxiv.org/abs/1705.09786). We replicate the MNIST results from the paper and add a benchmark on identity prediction LM task using transformers. Results are here: https://fb.quip.com/KPVkALAknU8i (FB internal).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by min-xu-ai** [comment]
quick question, should this be tagged experimental and somewhat organized separately from the "stable" codebase since it seems that it is still being validated on more models?

File: fairscale/nn/pipe/ampnet.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+                    optimizer.step()
+                    optimizer.zero_grad()
+
+                if vocab_size != 0:  # for real use-case
+                    word_counter += cur_batch["ntokens"]
+                    if count % log_interval == 0 and count > 0:
+                        total_loss += loss.item()
+                        cur_loss = total_loss / log_interval
+                        elapsed = time.time() - start_time
+                        print(
```
Comment (msbaines) [inline fairscale/nn/pipe/ampnet.py:270]: Use log instead of print.

**Review by msbaines — COMMENTED** [review]
Should we put amp net in experimental for now?

File: fairscale/nn/pipe/ampnet.py (hunk 1 above)
Comment (sidgoyal78) [reply fairscale/nn/pipe/ampnet.py:270]: Ohh, this was the old PR. I just created a new one: https://github.com/facebookresearch/fairscale/pull/304

**Comment by sidgoyal78** [comment]
Closing this one as it involves major refactor. See https://github.com/facebookresearch/fairscale/pull/304

**Pull request closed without merging**